### PR TITLE
ESQL: Improve pushdown of certain filters

### LINKED ISSUE #103536: ESQL:  IsNull/IsNotNull and other queries are not pushed down
### Description

The local physical planner is too strict and ends up not pushing down filters such as is null/is not null which heavily impact performance.

### PR DESCRIPTION
**Pull request by costin**
The check of what filters can be pushed down was too strict leading to
 some queries to be evaluated outside the storage. This PR addresses the
 issue in particular is not null and is null filters.

Fix #103536

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java [hunk 1]
```diff
@@ -240,10 +240,15 @@ public static boolean canPushToSource(Expression exp) {
             } else if (exp instanceof Not not) {
                 return canPushToSource(not.field());
             }
-            return false;
+            for (Attribute a : exp.references()) {
```
Comment (luigidellaquila) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java:243]: I'm afraid this is too generic.
It is problematic with boolean expressions that we are not considering in the previous IF, eg. functions that return a boolean

```
from employees | where to_boolean(to_string(emp_no))
```

```
{
    "error": {
        "root_cause": [
            {
                "type": "esql_illegal_argument_exception",
                "reason": "Expected a FieldAttribute or MetadataAttribute but received [to_boolean(to_string(emp_no))]"
            }
        ],
        "type": "esql_illegal_argument_exception",
        "reason": "Expected a FieldAttribute or MetadataAttribute but received [to_boolean(to_string(emp_no))]"
    },
    "status": 500
}
```

**Review by luigidellaquila — CHANGES_REQUESTED** [review]
I thing the checks on canPushToSource() have to be narrowed a bit. 
I left a comment with a problematic query

**Review by alex-spies — COMMENTED** [review]
The approach looks okay, I can see some test failures in CI though.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java]: Fixed.

**Review by astefan — APPROVED** [review]
LGTM

**Review by bpintea — APPROVED** [review]
Lgtm

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java [hunk 2]
```diff
@@ -101,15 +101,22 @@ public static Query toQuery(Expression e) {
...
 
     public static Query toQuery(Expression e, TranslatorHandler handler) {
+        Query translation = toQueryMaybe(e, handler);
+        if (translation != null) {
+            return translation;
+        }
+        throw new QlIllegalArgumentException("Don't know how to translate {} {}", e.nodeName(), e);
+    }
+
+    public static Query toQueryMaybe(Expression e, TranslatorHandler handler) {
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java]: Leftover - removed.

**Pull request merged**